recipes/elapid/meta.yaml

```diff
@@ -0,0 +1,50 @@
+{% set name = "elapid" %}
+{% set version = "0.3.13" %}
+
+package:
+  name: {{ name|lower }}
+  version: {{ version }}
+
+source:
+  url: "https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/{{ name }}-{{ version }}.tar.gz"
+  sha256: a713f02c5fcaec6cfe584c03cc1faed57dcb837b96d9bd8b21c479cd591164dd
+
+build:
+  skip: true  # [py<37]
+  script: {{ PYTHON }} -m pip install . -vv
+  number: 0
+
+requirements:
+  host:
+    - python
+    - pip
+  run:
```

Comment:
All these pins have no upper bound. I see that you are doing the same thing in the upstream package. You could consider being a bit more conservative here and bound things if you know/find out how their versioning works. Say, likely if there's ever a descartes 2.0.0 (I have no idea what that package does) then likely elapid 0.3.13 won't work with that version of descartes. So if people somehow try to install an old version of elapid, they'll get a new version of descartes (and everything else) and end up with a non-functional environment. On the other hand, if descartes 2.0.0 comes out and elapid works fine with it, then you can just relax the bound in a `build.number: 1`.
Anyway, just a suggestion.

Reply:
this is my first time preparing a distribution like this so I'm not familiar with best practices here. I was under the impression that I don't need to solve problems I don't yet have. That is, I don't need to anticipate breaking changes when nothing is yet broken. Is this naive?

I'm not sure I understand what setting the build number to 1 does. Does this mean that it's safer to restrict the major versions of most packages until they're released and tested to functionally work, at which point i'd change the requirement in the recipe to lift that restriction?

Reply:
reviewing the dependencies, it looks like there's only 1 or 2 packages that I might expect to introduce breaking changes, which are `geopandas` and `sklearn`.

`geopandas` seems unlikely to change with a 1.0 release, as they develop separate software systems that make fundamental changes to the api/dependencies, like [pyogrio](https://github.com/geopandas/pyogrio). I suspect the 1.0 release will still be backwards compatible but add features and increase stability.

`sklearn` has changed their API and deprecated features somewhat regularly, but have done so less frequently since the 1.0 release. I'm not sure how to best mitigate this - maybe bound it to <2.0?

The remaining dependencies are fairly stable (`numpy`, `rasterio`, `rtree`), are no longer in active development (`glmnet`), or breaking dependencies are handled upstream by `geopandas` (`descartes`, `shapely`).

So it doesn't seem super critical to pin upper bounds to me until we run into an issue, but I'm also willing to heed advice.

Reply:
I've at least pinned the upper version dependency of `glmnet` in case they resume development on the project, at which point I might expect some updates to their API.

Also, re-pinging @saraedum to kindly ask that they update their review.

Reply:
> this is my first time preparing a distribution like this so I'm not familiar with best practices here. I was under the impression that I don't need to solve problems I don't yet have. That is, I don't need to anticipate breaking changes when nothing is yet broken. Is this naive?

As I tried to outline above, having no upper bounds means that you might have to patch existing builds later. Say `descartes 2` breaks elapid, then somebody doing a `conda create -n something elapid=0.3.14` would get the latest version of `descartes` which won't work. So you need to patch the existing builds (which is somewhat cumbersome) to restrict `descartes <2` everywhere after the fact.

> I'm not sure I understand what setting the build number to 1 does. Does this mean that it's safer to restrict the major versions of most packages until they're released and tested to functionally work, at which point i'd change the requirement in the recipe to lift that restriction?

Yes.

Reply:
I made a more conservative suggestion below.

Feel free to ignore it if you don't like it. But it's easier to restrict things now than to react to breakage later imho.